### PR TITLE
Handle negative and zero sizes.

### LINKED ISSUE #11733: applyBoxFit doesn't properly handle an output size of Size.zero
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/box_fit.dart#L127

The result is a returned size with Nans in some cases resulting in issue: https://github.com/flutter/flutter/issues/11586

### PR DESCRIPTION
**Pull request by apwilson**
Fixes https://github.com/flutter/flutter/issues/11733

### CODE REVIEW

**Comment by abarth** [comment]
Looks reasonable, but we need a test.

**Comment by Hixie** [comment]
I think we should probably also audit the call sites. For example if we're trying to draw images at 0x0 we've probably messed up somewhere because it'd be much more efficient to not draw anything.

**Comment by apwilson** [comment]
I agree that the call sites should be audited but I don't think I'm equipped with the knowledge what to do in each case (there are three).

https://github.com/flutter/flutter/search?utf8=%E2%9C%93&q=applyBoxFit&type=

**Comment by apwilson** [comment]
And no, they don't handle zero size well but II'm sure they didn't handle Nan well either.

**Comment by Hixie** [comment]
flutter_logo.dart: On the line after we establish canvasSize, we should early-out if canvasSize.isEmpty.
box_painter.dart: paintImage should early-out on the line before it declares outputSize, if rect.isEmpty. Also, update the documentation of the `rect` argument to mention that if it's empty, nothing is painted.
proxy_box.dart: RenderFittedBox.hitTest should have the same early-out as RenderFittedBox.paint. RenderFittedBox.applyPaintTransform should have two branches. If size.isEmpty, it should call transform.setZero. The else branch should be everything that's there now.

**Comment by Hixie** [comment]
(Ideally you'd also have tests for the aforementioned changes, especially the changes to FittedBox.)

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 1]
```diff
@@ -1212,7 +1212,7 @@ Iterable<Rect> _generateImageTileRects(Rect outputRect, Rect fundamentalRect, Im
 ///  * `canvas`: The canvas onto which the image will be painted.
 ///  * `rect`: The region of the canvas into which the image will be painted.
 ///    The image might not fill the entire rectangle (e.g., depending on the
-///    `fit`).
+///    `fit`).  . If `rect` is empty, nothing is painted.
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/box_painter.dart]: Looks you you got an extraneous period.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**